app/models/concerns/foreman/observable_job.rb

```diff
@@ -0,0 +1,27 @@
+module Foreman
+  module ObservableJob
+    extend ActiveSupport::Concern
+    include Foreman::Observable
+
+    included do
+      class_attribute :event_subscription_hooks
+      self.event_subscription_hooks ||= []
```

Comment:
Does this work reliably with subclasses?

```ruby
class A < ApplicationJob
  include Foreman::ObservableJob
  set_hook :hook_a

class B < A
  include Foreman::ObservableJob
  set_hook :hook_b
end

B.perfrom_now
```

In this example, would performing `B` trigger only `hook_b` or `hook_a` as well?

Reply:
I'd expect `hook_b` only, I will add a test for that.

Reply:
So both B and A are triggered and I can't find a simple way to workaround that. Hmmm.

Reply:
what about something like

```ruby
class A < ApplicationJob
  include Foreman::ObservableJob
  set_hook :hook_a, if: -> { self.class == A }
end
```
?

Reply:
Well this requires devs to put the condition in, I can't make sure this will always be case.

I am wondering if it's safer to store hooks in hash of arrays where class is the hash key. Not sure how `self.class` evaluates in that context, need to test that.

Reply:
Guys, I've spent good amount of time with this and I don't know how to solve this keeping the current design. Major problem is that the moment the module is included, that's where the callback is registered. I was running into multiple problems with inheritance of both Rails callbacks and also concerns when I was trying various approaches. I am not able to figure out how to prevent triggering of hooks for the superclass.

Another problem I ran into was moving the common bit - `event_subscription_hooks` from model and job observer to the "common" observer class. There is some mixin/concern shadowing problem I could not figure out, so I am keeping the class attribute separate.

Lemme say this - it is good enough as long as we don't introduce inheritance like in @adamruzicka 's example above. Maybe if @kamils-iRonin want to solve this in another PR? My only idea is to create a singleton registry of Foreman hooks and define hooks for all `ApplicationRecord` and `ApplicationJob`. Then all processing and routing could be done in a single place. The only drawback would be that all records/jobs would be emitting callbacks even for those events which are non-subscribable.

my conclusion is - I don't want to tackle this problem here. A

Reply:
I looked into a bit more and I'd say the issue is we define an `after_perform` callback on the active job class at the moment when we call `set_hook` and active job callbacks get inherited when subclassing.

How about making `set_hook` only store all the needed arguments and then having a single `after_perform` which whould dispatch all of the stored hooks, instead of having one callback per hook?

Since it is quite an edge case, I don't mind not solving it here.

Reply:
Yes that is the direction I was thinking towards, although I did not think through all details. I've created https://projects.theforeman.org/issues/31646 for this, I really need to move on, work on packaging and then move to reports prototype which I started.